File: src/lib/analytics/__tests__/meal-timing.test.ts
import { describe, it, expect } from 'vitest';
import { extractMealTimingPatterns } from '../meal-timing';

describe('extractMealTimingPatterns', () => {
	it('computes basic eating window for a single day', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00+00:00', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T12:30:00+00:00', calories: 600 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T19:00:00+00:00', calories: 700 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows).toHaveLength(1);
		const day = result.dailyWindows[0];
		expect(day.firstMealTime).toBe('08:00');
		expect(day.lastMealTime).toBe('19:00');
		expect(day.windowMinutes).toBe(660);
		expect(day.mealCount).toBe(3);
		expect(day.lateNightMeals).toBe(0);
	});

	it('detects late-night meals after 21:00', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00+00:00', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T22:30:00+00:00', calories: 400 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T23:00:00+00:00', calories: 200 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows[0].lateNightMeals).toBe(2);
		expect(result.lateNightFrequency).toBe(100);
	});

	it('skips entries without eatenAt', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: null, calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T12:00:00+00:00', calories: 600 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows[0].mealCount).toBe(1);
	});

	it('computes correct hourly distribution', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T07:30:00+00:00', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T07:45:00+00:00', calories: 100 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T12:00:00+00:00', calories: 600 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T12:30:00+00:00', calories: 500 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.hourlyDistribution[7]).toBe(2);
		expect(result.hourlyDistribution[12]).toBe(2);
		expect(result.hourlyDistribution[0]).toBe(0);
		expect(result.hourlyDistribution).toHaveLength(24);
	});

	it('window is 0 for single meal per day', () => {
		const entries = [{ date: '2024-01-01', eatenAt: '2024-01-01T12:00:00+00:00', calories: 600 }];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows[0].windowMinutes).toBe(0);
		expect(result.dailyWindows[0].firstMealTime).toBe('12:00');
		expect(result.dailyWindows[0].lastMealTime).toBe('12:00');
	});

	it('buckets the UTC instant into the requested timezone (DST-correct)', () => {
		// 08:00Z and 18:00Z rendered in Europe/Zurich (winter, UTC+1) => 09:00, 19:00.
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00Z', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T18:00:00Z', calories: 600 }
		];
		const result = extractMealTimingPatterns(entries, 'Europe/Zurich');
		const day = result.dailyWindows[0];
		expect(day.firstMealTime).toBe('09:00');
		expect(day.lastMealTime).toBe('19:00');
	});

	it('returns empty summary for all-null eatenAt', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: null, calories: 300 },
			{ date: '2024-01-02', eatenAt: null, calories: 400 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows).toHaveLength(0);
		expect(result.avgWindowMinutes).toBe(0);
		expect(result.lateNightFrequency).toBe(0);
	});

	it('returns lateNightFrequency as percentage of days with late eating', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00+00:00', calories: 300 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T22:00:00+00:00', calories: 400 },
			{ date: '2024-01-03', eatenAt: '2024-01-03T12:00:00+00:00', calories: 500 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.lateNightFrequency).toBeCloseTo(33.33, 1);
	});

	it('resolves the same instant differently per timezone', () => {
		// 18:30Z is 00:00 next day in India (UTC+5:30) and 10:30 in PST (UTC-8).
		const entries = [{ date: '2024-01-01', eatenAt: '2024-01-01T18:30:00Z', calories: 500 }];
		expect(extractMealTimingPatterns(entries, 'Asia/Kolkata').dailyWindows[0].firstMealTime).toBe(
			'00:00'
		);
		expect(
			extractMealTimingPatterns(entries, 'America/Los_Angeles').dailyWindows[0].firstMealTime
		).toBe('10:30');
	});

	it('handles UTC (Z) timezone correctly', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T09:00:00Z', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T18:00:00Z', calories: 600 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows[0].firstMealTime).toBe('09:00');
		expect(result.dailyWindows[0].lastMealTime).toBe('18:00');
		expect(result.dailyWindows[0].windowMinutes).toBe(540);
	});

	it('normalizes mixed offset notations to the same instant', () => {
		// 09:00Z and 10:00+01:00 are the same instant; in UTC both bucket to 09:00.
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T09:00:00Z', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T10:00:00+01:00', calories: 600 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows).toHaveLength(1);
		expect(result.dailyWindows[0].firstMealTime).toBe('09:00');
		expect(result.dailyWindows[0].lastMealTime).toBe('09:00');
		expect(result.dailyWindows[0].mealCount).toBe(2);
	});

	it('handles timestamps with seconds and milliseconds', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T07:30:45.123+00:00', calories: 300 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T19:15:30.000+00:00', calories: 700 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows[0].firstMealTime).toBe('07:30');
		expect(result.dailyWindows[0].lastMealTime).toBe('19:15');
	});

	it('single entry per day yields window of 0 minutes', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T12:00:00Z', calories: 600 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T08:30:00Z', calories: 400 }
		];
		const result = extractMealTimingPatterns(entries, 'UTC');
		expect(result.dailyWindows).toHaveLength(2);
		for (const day of result.dailyWindows) {
			expect(day.windowMinutes).toBe(0);
			expect(day.firstMealTime).toBe(day.lastMealTime);
		}
	});
});

File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/MealTimingTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

class MealTimingTest {
    @Test
    fun localMinutesOfDayUtcZ() {
        val result = localMinutesOfDay("2024-01-15T08:30:00Z", "UTC")
        assertEquals(8 * 60 + 30, result)
    }

    @Test
    fun localMinutesOfDayConvertsToZone() {
        // 08:30Z rendered in Europe/Zurich (winter, UTC+1) => 09:30 => 570 minutes
        val result = localMinutesOfDay("2024-01-15T08:30:00Z", "Europe/Zurich")
        assertEquals(9 * 60 + 30, result)
    }

    @Test
    fun localMinutesOfDayBehindUtc() {
        // 08:30Z rendered in America/New_York (winter, UTC-5) => 03:30 => 210 minutes
        val result = localMinutesOfDay("2024-01-15T08:30:00Z", "America/New_York")
        assertEquals(3 * 60 + 30, result)
    }

    @Test
    fun localMinutesOfDayNormalizesOffsetToInstant() {
        // 08:30+02:00 is 06:30Z; in UTC that buckets to 06:30 => 390 minutes
        val result = localMinutesOfDay("2024-01-15T08:30:00+02:00", "UTC")
        assertEquals(6 * 60 + 30, result)
    }

    @Test
    fun localMinutesOfDayInvalidReturnsNull() {
        assertNull(localMinutesOfDay("not-a-timestamp", "UTC"))
        assertNull(localMinutesOfDay("2024-01-15", "UTC"))
        assertNull(localMinutesOfDay("", "UTC"))
    }

    @Test
    fun localMinutesOfDayInvalidZoneReturnsNull() {
        assertNull(localMinutesOfDay("2024-01-15T08:30:00Z", "Not/AZone"))
    }

    @Test
    fun localMinutesOfDayWrapsToNextLocalDay() {
        // 23:00Z rendered in Europe/Zurich (UTC+1) => 00:00 the next local day => 0 minutes
        val result = localMinutesOfDay("2024-01-15T23:00:00Z", "Europe/Zurich")
        assertEquals(0, result)
    }

    @Test
    fun extractMealTimingPatternsProducesCorrectWindow() {
        val entries =
            listOf(
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T07:00:00Z", calories = 400.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T12:30:00Z", calories = 600.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T19:00:00Z", calories = 700.0),
            )
        val result = extractMealTimingPatterns(entries, "UTC")
        assertEquals(1, result.dailyWindows.size)
        val window = result.dailyWindows[0]
        assertEquals("07:00", window.firstMealTime)
        assertEquals("19:00", window.lastMealTime)
        assertEquals(12 * 60, window.windowMinutes)
        assertEquals(3, window.mealCount)
        assertEquals(0, window.lateNightMeals)
    }

    @Test
    fun extractMealTimingPatternsBucketsIntoZone() {
        // Same instants as above shifted into Europe/Zurich (winter, UTC+1).
        val entries =
            listOf(
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T07:00:00Z", calories = 400.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T19:00:00Z", calories = 700.0),
            )
        val result = extractMealTimingPatterns(entries, "Europe/Zurich")
        val window = result.dailyWindows[0]
        assertEquals("08:00", window.firstMealTime)
        assertEquals("20:00", window.lastMealTime)
    }

    @Test
    fun extractMealTimingPatternsLateNightDetection() {
        val entries =
            listOf(
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T08:00:00Z", calories = 400.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T21:30:00Z", calories = 200.0),
            )
        val result = extractMealTimingPatterns(entries, "UTC")
        assertEquals(1, result.dailyWindows[0].lateNightMeals)
    }

    @Test
    fun extractMealTimingPatternsEmptyEntries() {
        val result = extractMealTimingPatterns(emptyList(), "UTC")
        assertEquals(emptyList(), result.dailyWindows)
        assertEquals(0.0, result.avgWindowMinutes)
        assertEquals("00:00", result.avgFirstMealTime)
        assertEquals("00:00", result.avgLastMealTime)
        assertEquals(0.0, result.lateNightFrequency)
    }

    @Test
    fun extractMealTimingPatternsHourlyDistribution() {
        val entries =
            listOf(
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T08:00:00Z", calories = 400.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T08:45:00Z", calories = 100.0),
                MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T12:00:00Z", calories = 600.0),
            )
        val result = extractMealTimingPatterns(entries, "UTC")
        assertEquals(2, result.hourlyDistribution[8])
        assertEquals(1, result.hourlyDistribution[12])
        assertEquals(0, result.hourlyDistribution[9])
    }

    @Test
    fun singleMealEntryProducesOneWindow() {
        val entries = listOf(MealEntry(date = "2024-01-01", eatenAt = "2024-01-01T12:00:00Z", calories = 500.0))
        val result = extractMealTimingPatterns(entries, "UTC")
        assertEquals(1, result.dailyWindows.size)
        assertEquals(1, result.dailyWindows[0].mealCount)
        assertEquals(0, result.dailyWindows[0].windowMinutes)
    }

    @Test
    fun allNullEatenAtReturnsEmpty() {
        val entries =
            listOf(
                MealEntry(date = "2024-01-01", eatenAt = null, calories = 500.0),
                MealEntry(date = "2024-01-02", eatenAt = null, calories = 400.0),
            )
        val result = extractMealTimingPatterns(entries, "UTC")
        assertEquals(0, result.dailyWindows.size)
    }

    @Test
    fun localMinutesOfDayMidnightExact() {
        val result = localMinutesOfDay("2024-01-01T00:00:00Z", "UTC")
        assertEquals(0, result)
    }

    @Test
    fun localMinutesOfDayEndOfDay() {
        val result = localMinutesOfDay("2024-01-01T23:59:00Z", "UTC")
        assertEquals(23 * 60 + 59, result)
    }

    @Test
    fun localMinutesOfDayWithSeconds() {
        val result = localMinutesOfDay("2024-01-01T08:30:45Z", "UTC")
        assertEquals(8 * 60 + 30, result)
    }
}
